lib/IRGen/GenOpaque.cpp

```diff
@@ -629,6 +629,82 @@ void IRGenFunction::emitDeallocateDynamicAlloca(StackAddress address) {
   Builder.CreateCall(stackRestoreFn, savedSP);
 }
 
+llvm::Value *IRGenFunction::emitStackAlloc(llvm::Value *size,
+                                           llvm::Value *align) {
+  auto returnType = IGM.Int8Ty;
+  auto returnPtrType = returnType->getPointerTo();
+  StackAddress stackPointer;
+  auto nullPointer = llvm::Constant::getNullValue(returnPtrType);
+
+  // Check that this function type can emit a stack allocation safely.
+  if (isAsync() || isCoroutine()) {
+    // These types of functions don't actually stack-allocate, and support for
+    // them is not needed at this time.
+    return nullPointer;
+  }
+
+  // Check that the caller-supplied alignment is usable.
+  Alignment resolvedAlign;
+  if (auto alignConst = dyn_cast<llvm::ConstantInt>(align)) {
+    // Ensure the supplied alignment is itself a power of 2.
+    auto alignConstValue = alignConst->getSExtValue();
+    if (!llvm::isPowerOf2_64(alignConstValue)) {
+      alignConstValue = llvm::NextPowerOf2(alignConstValue);
+    }
+
+    if (alignConstValue > 0) {
+      resolvedAlign = Alignment(alignConstValue);
+
+    } else {
+      // The specified alignment is nonsensical (negative, zero, or too big.)
+      return nullPointer;
+    }
+
+  } else {
+    // The caller specified a non-constant alignment value.
+    return nullPointer;
```

Comment:
The only valid use of this builtin is to support the proposed new stdlib functions. There's no real mechanism to propagate an error there (since failures imply a fallback to the heap), so I'm not certain what the benefit of an error would be?

Are you suggesting we could emit a diagnostic message if inputs are known-bad? That's reasonable (and in an earlier pass I did have some, but I took them out pending core team feedback.)

Reply:
See my reply in the other thread. :)